torch_xla/core/xla_model.py

```diff
@@ -974,6 +976,56 @@ def wait_device_ops(devices=[]):
   torch_xla._XLAC._xla_wait_device_ops(devices=devices)
 
 
+def bucketed_allreduce(gradients):
+    total = 0
+    tensor_bucket = []
+
+    for grad in gradients:
+      grad_bytes = grad.numel() * grad.element_size()
+
+      # Gradient is larger than bucket_cap, don't bucketize
+      if grad_bytes > bucket_cap:
+        # Flush out previous buckets even if they don't fill up
+        # This maintains the strict reverse ordering
+        if len(tensor_bucket):
+          all_reduce(
+              REDUCE_SUM,
+              tensor_bucket,
+              scale=1.0 / count,
+              groups=groups,
+              pin_layout=pin_layout)
+          total = 0
+          tensor_bucket = []
+        all_reduce(
+            REDUCE_SUM, [grad],
+            scale=1.0 / count,
+            groups=groups,
+            pin_layout=pin_layout)
+        continue
+
+      # Bucketize till the total spills over
+      total += grad_bytes
+      if total > bucket_cap:
+        all_reduce(
```

Comment:
Need to check "if len(tensor_bucket):" because tensor_bucket can be empty at the start, when grad_bytes > bucket_cap.